client/protocols/guidance.py

```diff
@@ -0,0 +1,178 @@
+            self.tg_id = self.debug
+        else:
+            self.get_tg(access_token,guidance_flag)
+
+        if self.tg_id == -1: 
+            # If self.tg_id == -1, some errors happen when trying to access server
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        lambda_string_key = self.guidance_json['dictTg2Func'].get(str(self.tg_id))
+
+        if not lambda_string_key:
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        lambda_info_misu = lambda_string_key_to_func.get(lambda_string_key)
+
+        lst_mis_u = dict_info.get('lstMisU')
+
+        # No list of misunderstandings for this wrong answer, default message
+        if not lst_mis_u:
+            print ("-- Not quite. Try again! --")
+            return ({}, -1,"")
+
+        self.misU_count = self.update_misUcounts(hash_key, lst_mis_u, repr(input_lines),shorten_unique_id)
+
+        thresold = self.guidance_json['wrongAnsThresh']
+
+        """
+        Update count file using the helper. Each misunderstanding igets one count.
+
+        Get the updated acculamted counts as a dictionary: key: mis_u, value: counts.
+        
+        Called that dictionary self.misU_count
+        """
+
+        self.msg_id_set = set()
+
+        # If the count is higher than the thresold, we need to display the message
+        for mis_u in self.misU_count:
+            if self.misU_count[mis_u] >= thresold:
+                # Print each associated unique message for each miunderstanding
+                self.msg_id = lambda_info_misu(dict_info, mis_u)
+
+                if self.msg_id:
+                    self.msg_id_set.add(self.msg_id)
+
+        if len(self.msg_id_set) == 0 or None in self.msg_id_set:
+            # if student is in control group, just print the default message
+            print ("-- Not quite. Try again! --")
+            return (self.misU_count, self.tg_id,"")
+
+        print("-- Helpful Hint --\n")
+
+        printed_out_msgs = ""
+        for mid in self.msg_id_set:
+            msg = self.guidance_json['dictId2Msg'][str(mid)]
+            printed_out_msgs = printed_out_msgs + msg
+            print(msg)
+
+        print("\n-- Helpful Hint --")
+
+        return (self.misU_count, self.tg_id,printed_out_msgs)
+
+    # Looks at the locally saved file for number of misU and returns the current count
+    def update_misUcounts(self, hashkey, misU, wrongAnswer, shorten_unique_id):
+        #Creates a new folder inside tests that stores the number of misU per assignment
+        count_file_path = "tests/misUcount.json"
+        if os.path.isfile(self.working_dir + count_file_path):
+            with open(self.working_dir + count_file_path,'r') as f:
+                jsonDic = json.load(f)
+                answerDict = jsonDic["answerDict"]
+                countData = jsonDic["countData"] 
+        else:
+            countData = {}
+            answerDict = {}
+        newjsonDic = {}
```

Comment:
This is not used until line 149, please move it closer to where it is first used.